### DISCUSSION TITLE
LLM Isnights (rate limit handling and so on )

### DISCUSSION THREAD

**Discussion by mdarweash** [langchain4j/langchain4j#881] [Q&A]
Does it support rate limit handling (like backoff, Retry) or is it on the path of being implemented?

any sort of collecting insights about LLM usage (number of tokens used ...etc) for later push to analytics (like prometheus or any data store)

**Comment by kwilliams-halosight** [comment]
I have another similar question, but I can answer in part. If your AI Service returns a Response<AIMessage> , the object has a Usage property object that contains token usage separated by input/output/etc.
